python/rainbow/simulators/prox_soft_bodies/collision_detection.py

```diff
@@ -38,6 +40,65 @@ def _update_bvh(engine, stats, debug_on):
     return stats
 
 
+def _spatial_hashing_narrow_phase(engine, stats, debug_on):
+    """ Use spatial hashing to find the overlapping triangles
+
+    Args:
+        engine (Engine): The current engine instance we are working with.
+        stats (dict): A dictionary where to add more profiling and timing measurements.
+        debug_on (bool): Boolean flag for toggling debug (aka profiling) info on and off.
+
+    Returns:
+        (List, dict):  A tuple with body pair overlap information and a dictionary with profiling and
+                        timing measurements.
+    """
+    narrow_phase_timer = None
+    if debug_on:
+        narrow_phase_timer = Timer("narrow_phase", 8)
+        narrow_phase_timer.start()
+    
+    cell_size = engine.hash_grid.cell_size
+    if cell_size <= 0.0:
+        raise ValueError("Cell size must be greater than zero")
+    
+    time_stamp = engine.params.time_stamp
+    results = defaultdict(set)
+
+    for body in engine.bodies.values():
+        tri_vertices = body.x[body.surface, :]
+        # Compute the AABB of each triangle by vectorizing the min/max operation
+        tri_aabb_min = np.min(tri_vertices, axis=1)
+        tri_aabb_max = np.max(tri_vertices, axis=1)
+        cell_min = (tri_aabb_min / cell_size).astype(int)
+        cell_max = (tri_aabb_max / cell_size).astype(int) + 1
+
+        # Traverse the cells in the AABB of each triangle and insert the triangle into the hash table
+        for tri_idx, (c_min, c_max) in enumerate(zip(cell_min, cell_max)):
+            cell_ranges = [range(cmi, cma) for cmi, cma in zip(c_min, c_max)]
+            for i, j, k in product(*cell_ranges):
+                tri_aabb = AABB(tri_aabb_min[tri_idx], tri_aabb_max[tri_idx])
+                overlaps = engine.hash_grid.insert(i, j, k, tri_idx, body.name, 
+                                                   tri_aabb, 
+                                                   time_stamp)
+                # Check all triangles in the cell to see if they overlap with the current triangle
+                if len(overlaps) > 0:
+                    for overlap in overlaps:
+                        overlap_tri_idx, overlap_body_name, overlap_tri_aabb = overlap
+                        if overlap_body_name != body.name and (AABB.is_overlap(tri_aabb, overlap_tri_aabb)):
```

Comment:
Hmm so self collision is not supported?

Test could be improved to add overlapping triangle pairs that do not share vertices that way self collision is supported for free?

Reply:
Self-collision will be supported in the spatial hashing approach, but I also found that there is not self-collision supported with BVH approach.

Reply:
Yes we never implemented this with BVHs. They can do it but one needs to have very good pruning rules to avoid descending too much. By its very definition a triangle will collide with it self so one basically ends up testing everything. The hashing approach does not have that feature. Here there are no difference between collision and self-collision.